src/parser/desugar.rs

```diff
@@ -0,0 +1,116 @@
+            Ast::variable(curried_arg, arg.locate())
+        } else {
+            arg
+        }
+    });
+
+    let function_call = Ast::function_call(function, args, loc);
+    Ast::lambda(curried_args, None, function_call, loc)
+}
+
+fn matches_underscore(arg: &Ast) -> bool {
+    matches!(arg, Ast::Variable(ast::Variable{ kind: ast::VariableKind::Identifier(x), ..}) if x == "_")
+}
+
+fn matches_not_typeconstructor(function: &Ast) -> bool {
+    !matches!(function, Ast::Variable(ast::Variable{ kind: ast::VariableKind::TypeConstructor(_), ..}))
+}
+
+/// Turns `2 |> foo _` into `foo 2` or `bar |> foo` into `foo bar`
+pub fn desugar_apply_operator<'a>(operator: Token, lhs: Ast<'a>, rhs: Ast<'a>, location: Location<'a>) -> Ast<'a> {
+    match operator {
+        Token::ApplyLeft  => prepend_argument_to_function(lhs, rhs, location),
+        Token::ApplyRight => prepend_argument_to_function(rhs, lhs, location),
+        _ => {
+            let operator = Ast::operator(operator, location);
+            Ast::function_call(operator, vec![lhs, rhs], location)
+        }
+    }
+}
+
+fn prepend_argument_to_function<'a>(f: Ast<'a>, arg: Ast<'a>, location: Location<'a>) -> Ast<'a> {
+    match f {
+        Ast::FunctionCall(mut call) => {
+            call.args.insert(0, arg);
+            Ast::FunctionCall(call)
+        },
+        Ast::Lambda(lambda @ ast::Lambda{..}) => {
+            apply_into_lambda(lambda, arg, location)
+        },
+        _ => Ast::function_call(f, vec![arg], location)
+    }
+}
+
+fn apply_into_lambda<'a>(lambda: ast::Lambda<'a>, new_arg: Ast<'a>, location: Location<'a>) -> Ast<'a> {
+    let mut lambda = lambda;
+
+    if let Ast::FunctionCall(ast::FunctionCall{..}) = *lambda.body{
+        if matches!(new_arg, Ast::Literal(_) | Ast::Variable(_)) {
+            if let Ast::FunctionCall(mut fcall @ ast::FunctionCall{..}) = *lambda.body {
+                // Lambdas cant have 0 arguments
+                if lambda.args.len() > 1 {
+                    let arg_to_replace =  lambda.args.remove(0);
+                    replace_lambda_body(&mut fcall, new_arg, arg_to_replace);
+                    return Ast::lambda(lambda.args, None, Ast::FunctionCall(fcall), location)
+                } else {
+                    let arg_to_replace =  lambda.args.remove(0);
+                    replace_lambda_body(&mut fcall, new_arg, arg_to_replace);
+                    return Ast::FunctionCall(fcall)
+                }
+            }
+        }
+    }
+
+    Ast::function_call(Ast::Lambda(lambda), vec![new_arg], location)
+}
+
+fn replace_lambda_body<'a>(f: &mut ast::FunctionCall<'a>, new_arg: Ast<'a>, arg_replace: Ast<'a>) {
+    let idxs: Vec<usize> = f.args
+                .iter()
+                .enumerate()
+                .filter(|(_, x)| name_matches(&arg_replace, x))
+                .map(|(i, _)| i).collect();
+
+    for idx in idxs {
+        match &new_arg {
+            Ast::Literal(lit) => {
```

Comment:
Honestly, I'm having second thoughts with this optimization. Applying it to all lambdas complicates things somewhat and makes me think that this is maybe better left to the optimizer. I thought it might be salvagable by changing the check to see if the argument is only used once instead, but that could still change evaluation order of something like `(print 1; 5) |> foo (print 2; 10) _`.
I think it'd be simpler if we just kept it as immediately calling the lambda so `3 |> print _` = `(fn $1 -> print $1) 3`

Reply:
I do agree, I think the potential savings down the line in terms of time to optimize isn't worth the complexity. What if i roll back the slotting and keep the seperate `desugar.rs` file?